### PR TITLE
Rewrite downloading the nuget dependencies.

### PR DESCRIPTION
**Pull request by annakocheshkova**
Things to consider before you submit the PR:

* [X] Are the files formatted correctly?
* [X] Did you test your change?

## Description

Rewrite downloading the nuget dependencies.

## Related PRs or issues

[AB#77174](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/77174)

### CODE REVIEW

File: build.cake [hunk 1]
```diff
@@ -378,43 +365,22 @@ Task("Install-Unity-Windows").Does(() => {
     }
 }).OnError(HandleError);
 
-// Downloading UWP IL2CPP dependencies.
+//Downloading UWP IL2CPP dependencies.
```
Comment (MatkovIvan) [inline build.cake]:
```suggestion
// Downloading UWP IL2CPP dependencies.
```

File: build.cake [hunk 2]
```diff
@@ -554,41 +520,26 @@ Task("DownloadNdk")
     }
 }).OnError(HandleError);
 
-void GetUwpPackage (AppCenterModule module, bool usePublicFeed) {
+async Task GetUwpPackage (AppCenterModule module, bool usePublicFeed)
```
Comment (MatkovIvan) [inline build.cake]:
```suggestion
async Task GetUwpPackage(AppCenterModule module, bool usePublicFeed) 
```

File: build.cake [hunk 3]
```diff
@@ -554,41 +520,26 @@ Task("DownloadNdk")
...
-            ExtractNuGetPackage (depPackage, new FrameworkName ("UAP, Version=v10.0"), nupkgPath, destination);
-            // Delete the package
-            DeleteFiles (nupkgPath);
-        }
-    } else {
-        nupkgPath = GetNuGetPackage (module.DotNetModule, UwpSdkVersion);
+    if (usePublicFeed) 
+    {
+        await ProcessDependency(new NugetDependency(module.DotNetModule, UwpSdkVersion, "UAP10.0"), destination);
+    } else
```
Comment (MatkovIvan) [inline build.cake]:
```suggestion
    }
    else 
```

File: build.cake [hunk 4]
```diff
@@ -554,41 +520,26 @@ Task("DownloadNdk")
...
+    {
+        nupkgPath = GetNuGetPackage(module.DotNetModule, UwpSdkVersion);
         var tempContentPath = "externals/uwp/" + module.Moniker + "/";
         DeleteDirectoryIfExists (tempContentPath);
-        // Unzip into externals/uwp/
         Unzip (nupkgPath, tempContentPath);
-        ExtractNuGetPackage (null, new FrameworkName ("UAP, Version=v10.0"), tempContentPath, destination);
-        // Delete the package
-        DeleteFiles (nupkgPath);
+        ProcessAppCenterDlls (tempContentPath, destination);
```
Comment (MatkovIvan) [inline build.cake]:
```suggestion
        ProcessAppCenterDlls(tempContentPath, destination);
```

File: build.cake [hunk 5]
```diff
@@ -554,41 +520,26 @@ Task("DownloadNdk")
...
     }
+    DeleteFiles (nupkgPath);
```
Comment (MatkovIvan) [inline build.cake]:
```suggestion
    DeleteFiles(nupkgPath);
```

File: nuget-tools.cake [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+string GetNuGetPackage(string packageId, string packageVersion)
+{
+    var nugetUser = EnvironmentVariable("NUGET_USER");
+    var nugetPassword = Argument("NuGetPassword", EnvironmentVariable("NUGET_PASSWORD"));
+    var nugetFeedId = Argument("NuGetFeedId", EnvironmentVariable("NUGET_FEED_ID"));
+    packageId = packageId.ToLower();
+    var filename = packageId + "." + packageVersion +  ".nupkg";
+    var url = $"{PrivateNugetFeedUrl}{nugetFeedId}/nuget/v3/flat2/{packageId}/{packageVersion}/{filename}";
+
+    // Get the NuGet package
```
Comment (MatkovIvan) [inline nuget-tools.cake]:
```suggestion
    // Get the NuGet package.
```

File: nuget-tools.cake [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            DeleteFile(targetPath);
+        }
+        MoveFile(matchingFile.FullPath, targetPath);
+        Information($"Moving {matchingFile.FullPath} to {targetPath}");
+    } 
+}
+
+void ProcessPackageDlls(NugetDependency package, string tempContentPath, string destination)
+{
+    Information($"ProcessPackageDlls; tempcontentpath {tempContentPath}; destination {destination}");
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Again, please avoid internal method names in the logs

File: build.cake [hunk 8]
```diff
@@ -1,20 +1,24 @@
...
+// Licensed under the MIT license.
+
 #addin nuget:?package=Cake.FileHelpers
 #addin nuget:?package=Cake.AzureStorage
-#addin nuget:?package=NuGet.Core
 #addin nuget:?package=Cake.Xcode
 #load "utility.cake"
+#load "nuget-tools.cake"
 
+using Path =  System.IO.Path;
```
Comment (MatkovIvan) [inline build.cake]:
```suggestion
using Path = System.IO.Path;
```

File: build.cake [hunk 9]
```diff
@@ -772,103 +723,10 @@ Task("clean")
...
-        }
-    }
-}
-
 void BuildXcodeProject(string projectPath)
 {
-    var projectFolder = System.IO.Path.GetDirectoryName(projectPath);
-    var buildOutputFolder =  System.IO.Path.Combine(projectFolder, "build");
+    var projectFolder = Path.GetDirectoryName(projectPath);
+    var buildOutputFolder =  Path.Combine(projectFolder, "build");
```
Comment (MatkovIvan) [inline build.cake]:
```suggestion
    var buildOutputFolder = Path.Combine(projectFolder, "build");
```

File: nuget-tools.cake [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Frameworks;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+using Path = System.IO.Path;
+
+var PrivateNugetFeedUrl = "https://msmobilecenter.pkgs.visualstudio.com/_packaging/";
```
Comment (MatkovIvan) [inline nuget-tools.cake]: const?

Comment (MatkovIvan) [reply nuget-tools.cake]: Revisit other similar places

File: nuget-tools.cake [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+void MoveNativeBinaries(string tempContentPath, string destination)
+{
+    var runtimesPath = tempContentPath + "/runtimes";
+    if (!DirectoryExists(runtimesPath)) 
+    {
+        return;
+    }
+    Information($"MoveNativeBinaries from {tempContentPath} to {destination}");
+    foreach (var runtime in GetDirectories(runtimesPath + "/win10-*")) {
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Format

File: nuget-tools.cake [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    MoveNativeBinaries(tempContentPath, destination);
+    var contentPathSuffix = "lib/uap10.0/";
+    var filesMask = tempContentPath + contentPathSuffix + '*';
+    Information($"Moving SDK package, move from {filesMask} to {destination}");
+    var files = GetFiles(filesMask);
+    CleanDirectories(destination);
+    MoveFiles(files, destination);
+}
+
+class PackageExtractor
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Why do we need a class with only static methods?

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: build.cake (hunk 8 above)
Comment (MatkovIvan) [reply build.cake]: Not resolved

File: nuget-tools.cake [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    var tempPackageFolder = ExternalsFolder + dependency.Name;
+    ExtractPackage(path, tempPackageFolder);
+    ProcessPackageDlls(dependency, tempPackageFolder, destination);
+    DeleteFiles(ExternalsFolder);
+}
+
+void ProcessInternalAppCenterDependency(string module, string moniker, string version, string destination) 
+{
+    var nupkgPath = GetPrivateNuGetPackage(module, UwpSdkVersion);
+    var tempContentPath = "externals/uwp/" + moniker + "/";
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Use ExternalsFolder

File: nuget-tools.cake [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+    MoveDllsByArchitecture(tempContentPath, destination);
+    var contentPathSuffix = "lib/uap10.0/";
+    var filesMask = tempContentPath + contentPathSuffix + '*';
+    Information($"Moving SDK package, move from {filesMask} to {destination}");
+    var files = GetFiles(filesMask);
+    CleanDirectories(destination);
+    MoveFiles(files, destination);
+}
+
+private void ExtractPackage(string fileName, string targetPath)
```
Comment (MatkovIvan) [inline nuget-tools.cake]: BTW, why not just unzip it? it's temporary data anyway

File: nuget-tools.cake [hunk 15]
```diff
@@ -0,0 +1,224 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#addin nuget:?package=NuGet.Resolver&loaddependencies=true
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Do we really need this now?

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: utility.cake [hunk 16]
```diff
@@ -66,7 +66,7 @@ static int ExecuteUnityCommand(string extraArgs, string projectPath = ".")
         {
             unityProcess.WaitForExit();
             result = unityProcess.GetExitCode();
-            if (logProcess.WaitForExit(0) && (logProcess.GetExitCode() != 0))
+            if (logProcess.WaitForExit(1000) && (logProcess.GetExitCode() != 0))
```
Comment (guperrot) [inline utility.cake]: What is this change?

Comment (MatkovIvan) [reply utility.cake]: It's picked from my branch. The problem was - the log process didn't contain the last error. It was killed to fast. In addition we delete log file, so the error message was completely missed even local (and for sure on CI)

Comment (MatkovIvan) [reply utility.cake]: https://github.com/microsoft/appcenter-sdk-unity/commit/8d5f8ce5c455f8213017ca6b5d1e02cdf3c63c61

Comment (MatkovIvan) [reply utility.cake]: @annakocheshkova please exclude unrelated change from this PR. Let's fix this separately

File: nuget-tools.cake (hunk 14 above)
Comment (russelarms) [reply nuget-tools.cake]: You're right.

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**